src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,158 @@
+using System.Diagnostics.Tracing;
+using System.Reflection;
+using System.Text.RegularExpressions;
+
+namespace OpenTelemetry.Instrumentation.EventCounters;
+
+/// <summary>
+/// .NET EventCounters Instrumentation.
+/// </summary>
+internal class EventCountersMetrics : EventListener
+{
+    internal static readonly Meter MeterInstance = new(AssemblyName.Name, AssemblyName.Version.ToString());
+    private static readonly AssemblyName AssemblyName = typeof(EventCountersMetrics).Assembly.GetName();
+    private static readonly Regex InstrumentNameRegex = new(
+        @"^[a-zA-Z][-.\w]{0,62}", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+
+    private readonly EventCountersInstrumentationOptions options;
+    private readonly ConcurrentQueue<EventSource> preInitEventSources = new();
+    private readonly ConcurrentDictionary<(string, string), Instrument> instruments = new();
+    private readonly ConcurrentDictionary<(string, string), double> values = new();
+    private readonly ConcurrentDictionary<string, byte> instrumentNames = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EventCountersMetrics"/> class.
+    /// </summary>
+    /// <param name="options">The options to define the metrics.</param>
+    public EventCountersMetrics(EventCountersInstrumentationOptions options)
+    {
+        this.options = options;
+
+        while (this.preInitEventSources.TryDequeue(out EventSource eventSource))
+        {
+            if (this.options.ShouldListenToSource(eventSource.Name))
+            {
+                this.EnableEvents(eventSource, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventSourceCreated(EventSource source)
+    {
+        if (this.options == null)
+        {
+            this.preInitEventSources.Enqueue(source);
+        }
+        else if (this.options.ShouldListenToSource(source.Name))
+        {
+            this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        if (this.options == null)
+        {
+            return;
+        }
+
+        var eventSourceName = eventData.EventSource.Name;
+
+        if (eventData.EventName != "EventCounters")
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreNonEventCountersName(eventSourceName);
+            return;
+        }
+
+        if (eventData.Payload == null || eventData.Payload.Count == 0 || eventData.Payload[0] is not IDictionary<string, object>)
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenEventArgsPayloadNotParseable(eventSourceName);
+            return;
+        }
+
+        var payload = eventData.Payload[0] as IDictionary<string, object>;
```

Comment:
nit: Update the `if` check above to cast it instead:
`.. || eventData.Payload[0] is not IDictionary<string, object> payload)`